### PR TITLE
🐛Fix Diagram Name Validation

### LINKED ISSUE #1142: 🐛 Aurelia Validation Does Not Work For "File with that name already exists" on Windows.
#### Describe your issue

When change the name of a diagram to a name that already exists a aurelia validation error should get displayed:

screenshot 2018-11-22 at 15 00 37

On windows this message this error does not get displayed, but when trying to finish the renaming the notification gets displayed.

screenshot 2018-11-22 at 15 00 55

Also the style of the validation error looks a bit broken.

This also causes #1291

#### How can others reproduce the issue?

**On windows:**

- Open up the BPMN-Studio 
- Open a solution
- Try to rename a diagram to a name that already exists
- **Notice that there is no validation error**

#### Issue checklist

Please check the boxes in this list after submitting your Issue:

- [x] I've checked if this issue already exists
- [x] I've included all the information that i think is relevant
- [x] I've added logs and/or screenshots (if applicable)
- [x] I've mentioned PRs and issues that relate to this one

### LINKED ISSUE #1291: 🐛 Windows: File Detection Does Not Work Anymore
#### Describe your issue

When adding a new file with an already existing name the existing file will be overwritten.

Related to #1142 

#### How can others reproduce the issue?

- Open up the BPMN-Studio 
- Open a solution
- Create a new file with the name of an already existing diagram
- **Notice that the file will be overwritten.**

#### Possible solution

~This might have something to do with this line in the solutionexplorer.repository.filesystem:~

~https://github.com/process-engine/solutionexplorer.repository.filesystem/pull/7/files#diff-04d298aa221f222670622a94f39f93f2R132~

#### Issue checklist

Please check the boxes in this list after submitting your Issue:

- [x] I've checked if this issue already exists
- [x] I've included all the information that i think is relevant
- [x] I've added logs and/or screenshots (if applicable)
- [x] I've mentioned PRs and issues that relate to this one

### PR DESCRIPTION
**Pull request by Vyperus**
**Changes:**

1. Add a refactored version of the diagram name validation.

**Issues:**

Closes #1142 
Closes #1291

PR: #PullRequest

## How can others test the changes?

Follow the instructions of the linked issue.

Notice that the faulty behavior wont occur on either Windows or macOS.

## PR-Checklist

Please check the boxes in this list after submitting your PR:

- [x] You can merge this PR **right now** (if not, please prefix the title with "WIP: ")
- [x] I've tested **all** changes included in this PR.
- [x] I've also reviewed this PR myself before submitting (e.g. for scrambled letters, typos, etc.).
- [x] I've rebased the `develop` branch with my branch before finishing this PR.
- [x] I've **summarized all changes** in a list above.
- [x] I've mentioned all **PRs, which relate to this one**.
- [x] I've prefixed my Pull Request title is according to [gitmoji guide](https://gitmoji.carloscuesta.me/).

### CODE REVIEW

File: src/modules/solution-explorer/solution-explorer-solution/solution-explorer-solution.ts [hunk 1]
```diff
@@ -839,4 +789,85 @@ export class SolutionExplorerSolution {
...
     }
   }
+
+  private _setValidationRules(): void {
+    ValidationRules
+      .ensure((state: IDiagramNameInputState) => state.currentDiagramInputValue)
+      .required()
+      .withMessage('Diagram name cannot be blank.')
+      .satisfies((input: string) => {
+        const inputNotIsEmpty: boolean = input !== undefined;
```
Comment (ElRaptorus) [inline src/modules/solution-explorer/solution-explorer-solution/solution-explorer-solution.ts]:
```suggestion
        const inputIsNotEmpty: boolean = input !== undefined;
```

File: src/modules/solution-explorer/solution-explorer-solution/solution-explorer-solution.ts [hunk 2]
```diff
@@ -839,4 +789,85 @@ export class SolutionExplorerSolution {
...
+
+        const inputAsCharArray: Array<string> = inputNotIsEmpty
```
Comment (ElRaptorus) [inline src/modules/solution-explorer/solution-explorer-solution/solution-explorer-solution.ts]:
```suggestion
        const inputAsCharArray: Array<string> = inputIsNotEmpty
```

File: src/modules/solution-explorer/solution-explorer-solution/solution-explorer-solution.ts [hunk 3]
```diff
@@ -839,4 +789,85 @@ export class SolutionExplorerSolution {
...
+          }
+
+          return true;
+        });
+
+        return diagramNamePassesNameChecks;
+      })
+      .withMessage(`Your diagram contains at least one invalid-character: \${$value}`)
+      .satisfies((input: string) => {
+        const inputNotIsEmpty: boolean = input !== undefined;
```
Comment (ElRaptorus) [inline src/modules/solution-explorer/solution-explorer-solution/solution-explorer-solution.ts]:
```suggestion
        const inputIsNotEmpty: boolean = input !== undefined;
```

File: src/modules/solution-explorer/solution-explorer-solution/solution-explorer-solution.ts [hunk 4]
```diff
@@ -839,4 +789,85 @@ export class SolutionExplorerSolution {
...
+
+        const diagramDoesNotStartWithWhitespace: boolean = inputNotIsEmpty
```
Comment (ElRaptorus) [inline src/modules/solution-explorer/solution-explorer-solution/solution-explorer-solution.ts]:
```suggestion
        const diagramDoesNotStartWithWhitespace: boolean = inputIsNotEmpty
```

File: src/modules/solution-explorer/solution-explorer-solution/solution-explorer-solution.ts [hunk 5]
```diff
@@ -839,4 +789,85 @@ export class SolutionExplorerSolution {
...
+                                                         ? !input.match(/^\s/)
+                                                         : true;
+
+        return diagramDoesNotStartWithWhitespace;
+      })
+      .withMessage('The diagram name can not start with a whitespace character.')
+      .satisfies((input: string) => {
+        const inputNotIsEmpty: boolean = input !== undefined;
```
Comment (ElRaptorus) [inline src/modules/solution-explorer/solution-explorer-solution/solution-explorer-solution.ts]:
```suggestion
        const inputIsNotEmpty: boolean = input !== undefined;
```

File: src/modules/solution-explorer/solution-explorer-solution/solution-explorer-solution.ts [hunk 6]
```diff
@@ -839,4 +789,85 @@ export class SolutionExplorerSolution {
...
+
+        const diagramDoesNotStartWithWhitespace: boolean = inputNotIsEmpty
+                                                         ? !input.match(/^\s/)
+                                                         : true;
+
+        return diagramDoesNotStartWithWhitespace;
+      })
+      .withMessage('The diagram name can not start with a whitespace character.')
```
Comment (ElRaptorus) [inline src/modules/solution-explorer/solution-explorer-solution/solution-explorer-solution.ts]:
```suggestion
      .withMessage('The diagram name cannot start with a whitespace character.')
```

File: src/modules/solution-explorer/solution-explorer-solution/solution-explorer-solution.ts [hunk 7]
```diff
@@ -839,4 +789,85 @@ export class SolutionExplorerSolution {
...
+      .satisfies((input: string) => {
+        const inputNotIsEmpty: boolean = input !== undefined;
+
+        const diagramDoesNotEndWithWhitespace: boolean = inputNotIsEmpty
```
Comment (ElRaptorus) [inline src/modules/solution-explorer/solution-explorer-solution/solution-explorer-solution.ts]:
```suggestion
        const diagramDoesNotEndWithWhitespace: boolean = inputIsNotEmpty
```

File: src/modules/solution-explorer/solution-explorer-solution/solution-explorer-solution.ts [hunk 8]
```diff
@@ -839,4 +789,85 @@ export class SolutionExplorerSolution {
...
+                                                       ? !input.match(/\s+$/)
+                                                       : true;
+
+        return diagramDoesNotEndWithWhitespace;
+      })
+      .withMessage('The diagram name can not end with a whitespace character.')
```
Comment (ElRaptorus) [inline src/modules/solution-explorer/solution-explorer-solution/solution-explorer-solution.ts]:
```suggestion
      .withMessage('The diagram name cannot end with a whitespace character.')
```

File: src/modules/solution-explorer/solution-explorer-solution/solution-explorer-solution.ts [hunk 9]
```diff
@@ -839,4 +789,85 @@ export class SolutionExplorerSolution {
...
+      .satisfies((input: string) => {
+        const inputNotIsEmpty: boolean = input !== undefined;
+
+        const inputAsCharArray: Array<string> = inputNotIsEmpty
+                                              ? input.split('')
+                                              : [];
+
+        const diagramNamePassesNameChecks: boolean = !inputAsCharArray.some((letter: string) => {
+          for (const regExIndex in this._diagramValidationRegExpList) {
+            const letterIsInvalid: boolean = letter.match(this._diagramValidationRegExpList[regExIndex]) !== null;
```
Comment (LeoTT) [inline src/modules/solution-explorer/solution-explorer-solution/solution-explorer-solution.ts]: I recommend `regex.test` instead of `string.match`. Returns directly `true` or `false`, is faster than match and you dont have to compare with `null`(which is error prone, because intuition would tell me to compare with `undefined`).

File: src/modules/solution-explorer/solution-explorer-solution/solution-explorer-solution.ts [hunk 10]
```diff
@@ -839,4 +789,85 @@ export class SolutionExplorerSolution {
...
+            if (letterIsInvalid) {
+              return false;
+            }
+          }
+
+          return true;
+        });
+
+        return diagramNamePassesNameChecks;
+      })
```
Comment (LeoTT) [inline src/modules/solution-explorer/solution-explorer-solution/solution-explorer-solution.ts]: I find the above code a bit hard to follow, because of negated some, multiple returns, etc.
How about:
```ts
const letterIsValid = (letter: string) => {
  const allRegExpChecksSucceeded = !this._diagramValidationRegExpList.some((regExp: RegExp) => {
    const validationFailed = regExp.test(letter) === false;
    return validationFailed;
  });
  return allRegExpsChecksSucceeded;
}
const diagramNamePassesNameChecks: boolean = inputAsCharArray.every(letterIsValid);
```

File: src/modules/solution-explorer/solution-explorer-solution/solution-explorer-solution.ts [hunk 11]
```diff
@@ -839,4 +789,80 @@ export class SolutionExplorerSolution {
...
+      .withMessage(`Your diagram contains at least one invalid-character: \${$value}`)
+      .satisfies((input: string) => {
+        const inputIsNotEmpty: boolean = input !== undefined;
+
+        const diagramDoesNotStartWithWhitespace: boolean = inputIsNotEmpty
+                                                         ? !input.match(/^\s/)
```
Comment (LeoTT) [inline src/modules/solution-explorer/solution-explorer-solution/solution-explorer-solution.ts]:
```suggestion
                                                         ? /^\s/.test(input) === false
```

File: src/modules/solution-explorer/solution-explorer-solution/solution-explorer-solution.ts [hunk 12]
```diff
@@ -839,4 +789,80 @@ export class SolutionExplorerSolution {
...
+                                                         : true;
+
+        return diagramDoesNotStartWithWhitespace;
+      })
+      .withMessage('The diagram name can not start with a whitespace character.')
+      .satisfies((input: string) => {
+        const inputIsNotEmpty: boolean = input !== undefined;
+
+        const diagramDoesNotEndWithWhitespace: boolean = inputIsNotEmpty
+                                                       ? !input.match(/\s+$/)
```
Comment (LeoTT) [inline src/modules/solution-explorer/solution-explorer-solution/solution-explorer-solution.ts]:
```suggestion
                                                       ? /\s+$/.test(input) === false
```
